pkg/console/complete.go

```diff
@@ -0,0 +1,245 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+package console
+
+import "strings"
+
+// CompleteLine returns a set of candidate TSQL keywords to complete the current input line
+func CompleteLine(line string) []string {
+	idx := strings.LastIndexAny(line, " ;") + 1
+	// we don't try to complete without a starting letter
+	if idx == len(line) {
+		return []string{}
+	}
+	prefix := strings.ToUpper(string(line[idx:]))
+	left := 0
+	right := len(keywords) - 1
+	for left <= right {
+		mid := (left + right) / 2
+		comp := 0
+		if len(keywords[mid]) >= len(prefix) {
+			comp = strings.Compare(prefix, string(keywords[mid][:len(prefix)]))
+		} else {
+			comp = strings.Compare(prefix, keywords[mid])
+		}
+		if comp < 0 {
+			right = mid - 1
+		} else if comp > 0 {
+			left = mid + 1
+		} else {
+			// look up and down the list from mid and return the slice of matching words
+			first := mid - 1
+			last := mid + 1
+			for first >= 0 && strings.HasPrefix(keywords[first], prefix) {
+				first--
+			}
+			for last < len(keywords) && strings.HasPrefix(keywords[last], prefix) {
+				last++
+			}
+			lines := make([]string, last-first-1)
+			for i, w := range keywords[first+1 : last] {
+				lines[i] = mergeLine(line, w, idx)
+			}
+			return lines
+		}
+	}
+	return []string{}
+}
+
+// mergeline appends keyword to line starting at index idx
+// It matches the case of the current character in the line
+func mergeLine(line string, keyword string, idx int) string {
```

Comment:
Question - should we add a space after the keyword? I find it natural to hit the spacebar myself after the tab cycles to the word I am looking for but maybe that's just me.

Reply:
any completion functionality I'm familiar with also does not add the space after the keyword, I do.  This is especially important for when you might accidentally complete to a keyword that is a component of another keyword.